cpp/src/arrow/filesystem/s3fs.cc

```diff
@@ -340,6 +340,12 @@ class ObjectInputFile : public io::RandomAccessFile {
     RETURN_NOT_OK(CheckClosed());
     RETURN_NOT_OK(CheckPosition(position, "read"));
 
+    nbytes = std::min(nbytes, content_length_ - position);
```

Comment:
@pitrou this is the fix for `stream.read()` issue

Reply:
Thanks! And thanks for the tests too.